File: Java/ps_baekjoon/14003/Main.java
//baekjoon ps 14003

import java.util.*;
import java.util.stream.*;
import java.io.*;

public class Main {
    static int INF = Integer.MAX_VALUE;

    static class Pair {
        int first, second;

        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = Stream.of(br.readLine().split(" ")).limit(n).mapToInt(Integer::parseInt).toArray();
        ArrayList<Integer> lis = new ArrayList<>();
        ArrayList<Integer> res = new ArrayList<>();
        ArrayList<Pair> trace = new ArrayList<>();
        lis.add(-INF);
        for (int i = 0; i < n; i++) {
            int data = arr[i];
            if (lis.get(lis.size() - 1) < data) {
                trace.add(new Pair(lis.size(), data));
                lis.add(data);
            } else {
                int idx = Collections.binarySearch(lis, data);
                idx = idx < 0 ? idx * (-1) - 1 : idx;
                lis.set(idx, data);
                trace.add(new Pair(idx, data));
            }
        }
        for (int i = trace.size() - 1, j = lis.size() - 1; i >= 0; i--)
            if (trace.get(i).first == j) {
                res.add(trace.get(i).second);
                j--;
            }
        Collections.reverse(res);
        System.out.println(res.size());
        res.stream().forEach((e) -> System.out.print(e + " "));
    }
}

File: C++/ps_baekjoon/14003/main.cpp
//baekjoon ps 14003

#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>
#include <limits>

using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    constexpr int INF = numeric_limits<int>::max();
    int n;
    cin >> n;
    vector<int> arr(n), lis, res;
    vector<pair<int, int>> trace;
    lis.push_back(-INF);
    for (int i = 0; i < n; i++)
        cin >> arr[i];
    for (int i = 0; i < n; i++)
    {
        int data = arr[i];
        if (lis.back() < data)
        {
            trace.push_back({lis.size(), data});
            lis.push_back(data);
        }
        else
        {
            auto iter = lower_bound(lis.begin(), lis.end(), data);
            *iter = data;
            trace.push_back({iter - lis.begin(), data});
        }
    }
    for (int i = trace.size() - 1, j = lis.size() - 1; i >= 0; i--)
        if (trace[i].first == j)
        {
            res.push_back(trace[i].second);
            j--;
        }
    reverse(res.begin(), res.end());
    cout << lis.size() - 1 << '\n';
    for (int i = 0; i < res.size(); i++)
        cout << res[i] << ' ';
    return 0;
}